traitlets/traitlets.py

```diff
@@ -418,6 +420,10 @@ def _dynamic_default_callable(self, obj):
         return getattr(self, 'make_dynamic_default', None)
 
     def instance_init(self, obj):
+        if not hasattr(obj, '_trait_values'):
```

Comment:
I talked to @ellisonbg and his sentiment was that this sort of logic should be kept on `HasDescriptor` in order to simplify writing the descriptors themselves.